rest-service/manager_rest/rest/filters_utils.py

```diff
@@ -1,129 +1,138 @@
-import re
+from typing import List
 
 from flask import request
 
-from manager_rest import manager_exceptions
+from cloudify._compat import text_type
+
 from manager_rest.rest.rest_utils import validate_inputs
+from manager_rest.manager_exceptions import BadFilterRule, BadParametersError
 from manager_rest.storage import get_storage_manager, models
-from manager_rest.constants import EQUAL, NOT_EQUAL, IS_NULL, IS_NOT_NULL
-
+from manager_rest.constants import (ATTRS_OPERATORS,
+                                    FilterRuleType,
+                                    LabelsOperator,
+                                    LABELS_OPERATORS,
+                                    FILTER_RULE_TYPES)
 
-class BadLabelsFilter(manager_exceptions.BadParametersError):
-    def __init__(self, labels_filter_value):
-        super(BadLabelsFilter, self).__init__(
-            'The labels filter `{0}` is not in the right format. It must be '
-            'one of: <key>=<value>, <key>=[<value1>,<value2>,...], '
-            '<key>!=<value>, <key>!=[<value1>,<value2>,...], <key> is null, '
-            '<key> is not null'.format(labels_filter_value)
-        )
 
-
-def get_filter_rules():
+def get_filter_rules(resource_model: models):
     filter_rules = request.args.get('_filter_rules')
     filter_id = request.args.get('_filter_id')
 
     if not filter_rules and not filter_id:
         return
 
     if filter_rules and filter_id:
-        raise manager_exceptions.BadParametersError(
-            'Filter rules and filter name cannot be provided together. '
+        raise BadParametersError(
+            'Filter rules and filter ID cannot be provided together. '
             'Please specify one of them or neither.'
         )
 
     if filter_rules:
-        return create_labels_filters_mapping(filter_rules.split(','))
+        return parse_filter_rules(filter_rules, resource_model)
 
     if filter_id:
         validate_inputs({'filter_id': filter_id})
         filter_elem = get_storage_manager().get(models.Filter, filter_id)
-        return filter_elem.value.get('labels', {})
-
-
-def create_labels_filters_mapping(labels_filters_list):
-    """Validate and parse a list of labels filters
-
-    :param labels_filters_list: A list of labels filters. Labels filters must
-           be one of: <key>=<value>, <key>=[<value1>,<value2>,...],
-           <key>!=<value>, <key>!=[<value1>,<value2>,...], <key> is null,
-           <key> is not null
-
-    :return The labels filters mapping dictionary with the following schema:
-            {NOT_EQUAL: {}, EQUAL: {}, IS_NULL: [], IS_NOT_NULL: []}
+        return filter_elem.value
+
+
+def parse_filter_rules(filter_rules: List[dict], resource_model: models):
+    """Validating the filter rules list.
```

Comment:
```suggestion
    """Validate the filter rules list.
```

but surely this doesn't just validate, it parses?

Reply:
Well, it mainly validates, the only parsing is converting the key and values to lowercase. But I'll modify it to only validate, and the lowercase change will be moved to the filter rule class.